cirq-core/cirq/transformers/analytical_decompositions/pauli_string_decomposition.py

```diff
@@ -0,0 +1,112 @@
+# Copyright 2023 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional, Tuple, cast
+
+import numpy as np
+import numpy.typing as npt
+
+from cirq.ops import DensePauliString
+from cirq import protocols
+
+
+def _argmax(V: npt.NDArray) -> Tuple[int, float]:
+    V = (V * V.conj()).real
+    idx_max = np.argmax(V)
+    V[idx_max] = 0
+    return cast(int, idx_max), np.max(V)
+
+
+def _validate_decomposition(decomposition: DensePauliString, U: npt.NDArray, eps: float) -> bool:
+    got = protocols.unitary(decomposition)
+    return np.abs(got - U).max() < eps
+
+
+def _fast_walsh_hadamard_transform(V: npt.NDArray) -> None:
+    """Fast Walsh–Hadamard Transform of an array."""
+    m = len(V)
+    n = m.bit_length() - 1
+    for h in [2**i for i in range(n)]:
+        for i in range(0, m, h * 2):
+            for j in range(i, i + h):
+                x = V[j]
+                y = V[j + h]
+                V[j] = x + y
+                V[j + h] = x - y
+
+
+def _conjugate_with_hadamard(U: npt.NDArray) -> npt.NDArray:
+    """Applies H†UH in O(n4^n) instead of O(8^n)."""
+
+    U = np.copy(U.T)
+    for i in range(U.shape[1]):
+        _fast_walsh_hadamard_transform(U[:, i])
+    U = U.T
+    for i in range(U.shape[1]):
+        _fast_walsh_hadamard_transform(U[:, i])
+    return U
+
+
+def unitary_to_pauli_string(U: npt.NDArray, eps: float = 1e-15) -> Optional[DensePauliString]:
+    """Attempts to find a pauli string (with possible phase) equivalent to U up to eps.
+
+        Based on this answer https://shorturl.at/aA079.
```

Comment:
I think it might be worth restating the algorithm here as well as linking to the source.  This is pretty hard to follow otherwise.

Maybe add some line comments below too.

Reply:
done